l1/l1.go

```diff
@@ -41,24 +43,46 @@ func NewClient(l1 Subscriber, chain *blockchain.Blockchain, log utils.SimpleLogg
 		l2Chain:          chain,
 		log:              log,
 		network:          chain.Network(),
+		resubscribeDelay: 10 * time.Second,
 		nonFinalisedLogs: make(map[uint64]*contract.StarknetLogStateUpdate, 0),
 	}
 }
 
+func (c *Client) WithResubscribeDelay(delay time.Duration) *Client {
+	c.resubscribeDelay = delay
+	return c
+}
+
 func (c *Client) subscribeToHeaders(ctx context.Context, headerChan chan *types.Header) (event.Subscription, error) {
-	headerSub, err := c.l1.WatchHeader(ctx, headerChan)
-	if err != nil {
-		return nil, fmt.Errorf("subscribe to L1 headers: %w", err)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context canceled before resubscribe was successful: %w", ctx.Err())
+		default:
+			headerSub, err := c.l1.WatchHeader(ctx, headerChan)
+			if err == nil {
+				return headerSub, nil
+			}
+			c.log.Warnw("Failed to subscribe to L1 headers", "tryAgainIn", c.resubscribeDelay, "err", err)
+			time.Sleep(c.resubscribeDelay)
+		}
```

Comment:
Do you think it is worth extracting this loop to a function to be reused in subscribeToUpdates

Reply:
Probably more effort than it's worth. The real solution is to drop `WatchHeader` altogether, since we just use it as a `time.Ticker`.